python/ray/air/examples/dreambooth/run_model.py

```diff
@@ -3,29 +3,76 @@
 
 from diffusers import DiffusionPipeline
 import torch
+import ray
 
 from flags import run_model_flags
 
 
 def run(args):
-    print(f"Loading model from {args.model_dir}")
-    pipeline = DiffusionPipeline.from_pretrained(
-        args.model_dir, torch_dtype=torch.float16
-    )
-    pipeline.set_progress_bar_config(disable=True)
-    if torch.cuda.is_available():
-        pipeline.to("cuda")
+    class StableDiffusionCallable:
+        def __init__(self, model_dir, output_dir):
+            print(f"Loading model from {model_dir}")
+
+            self.pipeline = DiffusionPipeline.from_pretrained(
+                model_dir, torch_dtype=torch.float16
+            )
+            self.pipeline.set_progress_bar_config(disable=True)
+            if torch.cuda.is_available():
+                self.pipeline.to("cuda")
+            self.output_dir = output_dir
+
+        def __call__(self, batch):
+            filenames = []
+            for i, prompt in zip(batch["idx"], batch["prompt"]):
+                # Generate 1 image at a time to reduce memory consumption.
+                for image in self.pipeline(prompt).images:
+                    hash_image = hashlib.sha1(image.tobytes()).hexdigest()
+                    image_filename = path.join(self.output_dir, f"{i}-{hash_image}.jpg")
+                    image.save(image_filename)
+                    print(f"Saved {image_filename}")
+                    filenames.append(image_filename)
+            return {"filename": filenames}
 
     prompts = args.prompts.split(",")
 
-    # Generate 1 image to reduce memory consumption.
-    for prompt in prompts:
-        for i in range(args.num_samples_per_prompt):
-            for image in pipeline(prompt).images:
-                hash_image = hashlib.sha1(image.tobytes()).hexdigest()
-                image_filename = path.join(args.output_dir, f"{i}-{hash_image}.jpg")
-                image.save(image_filename)
-                print(f"Saved {image_filename}")
+    if args.use_ray_data:
+        # Use Ray Data to perform batch inference to generate many images in parallel
```

Comment:
Is this the new recommended way to do batch inference now? It gives user more flexibility compared with BatchPredictor.

Reply:
Yes, the map batches w/ a callable!